File: Design-Patterns-Using-JavaScript/Iterator/IteratorPatternLinkedList.js
class Node {
    constructor(value, next) {
        this.value = value;
        this.next = next;
    }
}

class Iterator {
    next() {}
    hasNext() {}
}

class LinkedListIterator extends Iterator {
    constructor(aggregate) {
        super();
        this.aggregate = aggregate;
        this.current = aggregate.head;
    }

    hasNext() {
        return this.current !== null;
    }

    next() {
        if (!this.hasNext()) {
            throw new Error("No such element");
        }
        const value = this.current.value;
        this.current = this.current.next;
        return value;
    }
}

class Iterable {
    iterator() {}
}

class LinkedList extends Iterable {
    constructor() {
        super();
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    addHead(value) {
        const newNode = new Node(value, this.head);
        if (!this.head) {
            this.tail = newNode;
        }
        this.head = newNode;
        this.size++;
    }

    iterator() {
        return new LinkedListIterator(this);
    }

    getSize() {
        return this.size;
    }
}

// Client code
const linkedList = new LinkedList();
linkedList.addHead(1);
linkedList.addHead(2);
linkedList.addHead(3);

const iterator = linkedList.iterator();
while (iterator.hasNext()) {
    console.log(iterator.next());
}

/*
3
2
1
*/

File: Design-Patterns-Using-Kotlin/Iterator/IteratorPatternLinkedList.kt
class LinkedList : Iterable<Int?> {
    inner class Node(var value: Int, var next: Node?)

    var head: Node? = null
    var tail: Node? = null
    var size: Int = 0

    fun addHead(value: Int) {
        val newNode: Node = Node(value, head)
        if (head == null) {
            tail = newNode
        }
        head = newNode
        size++
    }

    override fun iterator(): Iterator<Int> {
        return LinkedListIterator(this)
    }
}

class LinkedListIterator(var aggregate: LinkedList) : Iterator<Int> {
    var current: LinkedList.Node?
    var previous: LinkedList.Node? = null

    init {
        this.current = aggregate.head
    }

    override fun hasNext(): Boolean {
        return current != null
    }

    override fun next(): Int {
        if (!hasNext()) {
            throw NoSuchElementException()
        }
        val value = current!!.value
        previous = current
        current = current!!.next
        return value
    }
}

// Client Code
fun main() {
    val aggregate = LinkedList()
    for (i in 0..4) {
        aggregate.addHead(i)
    }

    for (`val` in aggregate) {
        print("$`val` ")
    }
}

/*
4 3 2 1 0 
*/